zcash_client_backend/src/keys.rs

```diff
@@ -144,16 +195,151 @@ impl UnifiedSpendingKey {
+        {
+            let account_tkey = self.transparent();
+            CompactSize::write(&mut result, usize::from(Typecode::P2pkh)).unwrap();
+
+            let account_tkey_bytes = account_tkey.to_bytes();
+            CompactSize::write(&mut result, account_tkey_bytes.len()).unwrap();
+            result.write_all(&account_tkey_bytes).unwrap();
+        }
+
+        result
+    }
+
+    #[allow(clippy::unnecessary_unwrap)]
+    #[cfg(feature = "unstable")]
+    pub fn from_bytes(epoch: Epoch, encoded: &[u8]) -> Result<Self, DecodingError> {
+        let mut source = std::io::Cursor::new(encoded);
+        let decoded_epoch = source
+            .read_u32::<LittleEndian>()
+            .map_err(|_| DecodingError::ReadError("epoch"))
+            .and_then(|id| Epoch::from_id(id).ok_or(DecodingError::EpochInvalid))?;
+
+        if decoded_epoch != epoch {
+            return Err(DecodingError::EpochMismatch(decoded_epoch));
+        }
+
+        let mut orchard = None;
+        let mut sapling = None;
+        #[cfg(feature = "transparent-inputs")]
+        let mut transparent = None;
+        loop {
+            let tc = CompactSize::read_t::<_, u32>(&mut source)
+                .map_err(|_| DecodingError::ReadError("typecode"))
+                .and_then(|v| Typecode::try_from(v).map_err(|_| DecodingError::TypecodeInvalid))?;
+
+            let len = CompactSize::read_t::<_, u32>(&mut source)
+                .map_err(|_| DecodingError::ReadError("key length"))?;
+
+            match tc {
+                Typecode::Orchard => {
+                    if len != 32 {
+                        return Err(DecodingError::LengthMismatch(Typecode::Orchard, len));
+                    }
+
+                    let mut key = [0u8; 32];
+                    source
+                        .read_exact(&mut key)
+                        .map_err(|_| DecodingError::InsufficientData(Typecode::Orchard))?;
+                    orchard = Some(
+                        Option::<orchard::keys::SpendingKey>::from(
+                            orchard::keys::SpendingKey::from_bytes(key),
+                        )
+                        .ok_or(DecodingError::KeyDataInvalid(Typecode::Orchard))?,
+                    );
+                }
+                Typecode::Sapling => {
+                    if len != 169 {
+                        return Err(DecodingError::LengthMismatch(Typecode::Sapling, len));
+                    }
+
+                    let mut key = [0u8; 169];
+                    source
+                        .read_exact(&mut key)
+                        .map_err(|_| DecodingError::InsufficientData(Typecode::Sapling))?;
+                    sapling = Some(
+                        sapling::ExtendedSpendingKey::from_bytes(&key)
+                            .map_err(|_| DecodingError::KeyDataInvalid(Typecode::Sapling))?,
+                    );
+                }
+                #[cfg(feature = "transparent-inputs")]
```

Comment:
If the transparent feature is not enabled, should a USK that includes a P2PKH key still be accepted? Or do we require that all components communicating with each other are consistently compiled either to use or not use transparent (even across languages)?

Reply:
That's a good point; this is a situation where the use of a feature flag to control transparent functionality is a bit awkward. I guess that for parsing purposes, if transparent functionality is not enabled then we should raise an error if we are asked to parse a USK containing a transparent component.

Reply:
Given how we intend the USK to be used an error like this should probably be treated as a developer error, shouldn't it? I can't think of a users getting into this problem by just using the wallet because the USK is encoded in binary for internal interchange between the SDK/Wallet and the Rust crate. 

If I'm not mistaken is not possible to call `shield_funds` if the `transparent-inputs` is not enabled. There could be the case that a developer mistakenly saves the bytes somewhere and then if they move out of a build that includes such feature, they end up passing a USK with transparent keys. This is probably an error in migrating their software that was already using these keys as they are not intended to. :-/

I think an error must be raised to prevent this situation go unnoticed.

Reply:
Actually, if `feature = "transparent-inputs"` is not enabled, at present this *will* return a `TypecodeInvalid` error.

Reply:
Right. My question was, should the transparent part be ignored instead?

Reply:
Ah, okay. No, I think that this should be an error, because if the key was originally created with a transparent part, it may control transparent funds that will be otherwise unrecoverable. So I think an error is correct here.